gtest_parallel.py

```diff
@@ -402,27 +403,108 @@ def __init__(self, json_dump_filepath):
         "tests": {},
     }
 
-  def log(self, test, runtime_ms, actual_result):
+  def log(self, task):
+    actual_result = "PASS" if task.exit_code == 0 else "FAIL"
     with self.test_results_lock:
       self.test_results['num_failures_by_type'][actual_result] += 1
       results = self.test_results['tests']
-      for name in test.split('.'):
+      for name in task.test_name.split('.'):
         results = results.setdefault(name, {})
 
       if results:
         results['actual'] += ' ' + actual_result
-        results['times'].append(runtime_ms)
+        results['times'].append(task.runtime_ms)
       else:  # This is the first invocation of the test
         results['actual'] = actual_result
-        results['times'] = [runtime_ms]
-        results['time'] = runtime_ms
+        results['times'] = [task.runtime_ms]
+        results['time'] = task.runtime_ms
         results['expected'] = 'PASS'
 
   def dump_to_file_and_close(self):
     json.dump(self.test_results, self.json_dump_file)
     self.json_dump_file.close()
 
 
+class CollectXMLTestResults(object):
+  def __init__(self, xml_dump_filepath, all_logs):
+    self.test_results_lock = threading.Lock()
+    # Include in XML standard output of succeeded test
+    self.all_logs = all_logs
+    self.xml_dump_filepath = xml_dump_filepath
+    self.test_results = {
+      "tag": "testsuites",
+      "attrib": {
+        "name": "AllTests",
+        "disabled": 0,
+        "errors": 0,
+        "timestamp": datetime.datetime.now().isoformat(),
+        "time": 0,
+      },
+      "children": {},
+    }
+
+  def log(self, task):
+    failed = task.exit_code != 0
+    with self.test_results_lock:
+      task_name = task.test_name.split('.')
+      group_name = task_name[0]
+      test_name = '.'.join(task_name[1:])
+      group = self.test_results['children'].setdefault(group_name, {
+        "tag": "testsuite",
+        "attrib": {
+          "name": group_name,
+          "disabled": 0,
+          "errors": 0,
+        },
+        "children": {},
+      })
+      assert test_name not in group['children'], "Multiple runs are not supported with XML test results output"
```

Comment:
This needs to be verified earlier as part of command line validation (keep the assert though). See `parser.error` usage in `main()` for instance.

Reply:
Yes, failing earlier is better. I'm not sure what turns on multiple runs, I'll take a closer look.